doesn't work with mypy

```
$ python3.7 -m venv env
$ . env/bin/activate
(env) $ pip install sansio-lsp-client
...
Successfully installed attrs-18.2.0 cattrs-0.9.2 sansio-lsp-client-0.6.1
(env) $ pip install mypy
...
Successfully installed mypy-0.782 mypy-extensions-0.4.3 typed-ast-1.4.1 typing-extensions-3.7.4.2
(env) $ echo import sansio_lsp_client > hello.py
(env) $ mypy hello.py
hello.py:1: error: Skipping analyzing 'sansio_lsp_client': found module but no type hints or library stubs
hello.py:1: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 1 source file)
```

i'm a bit confused because there are type hints in `sansio_lsp_client`

--- a/.github/workflows/black.yml
+++ b/.github/workflows/black.yml
@@ -1,0 +1,19 @@
+# mostly copied from mypy.yml
+name: Black
+on: [push]
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    name: Black
+    steps:
+    - uses: actions/checkout@v1
+    - name: Set up Python 3.8
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.8
+    - name: Install Dependencies
+      run: |
+        pip install poetry && poetry install
+    - name: black
+      run: |
+        poetry run python3 -m black --check sansio_lsp_client

--- a/.github/workflows/black.yml
+++ b/.github/workflows/black.yml
@@ -1,6 +1,6 @@
 # mostly copied from mypy.yml
 name: Black
-on: [push]
+on: [push, pull_request]
 jobs:
   build:
     runs-on: ubuntu-latest

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -1,6 +1,6 @@
 # https://medium.com/@doedotdev/mypy-for-github-action-7da1ebee99e7
 name: Mypy
-on: [push]
+on: [push, pull_request]
 jobs:
   build:
     runs-on: ubuntu-latest

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -1,0 +1,19 @@
+# https://medium.com/@doedotdev/mypy-for-github-action-7da1ebee99e7
+name: Mypy
+on: [push]
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    name: Mypy
+    steps:
+    - uses: actions/checkout@v1
+    - name: Set up Python 3.8
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.8
+    - name: Install Dependencies
+      run: |
+        pip install mypy
+    - name: mypy
+      run: |
+        mypy --strict sansio_lsp_client

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -16,4 +16,4 @@ jobs:
         pip install poetry && poetry install
     - name: mypy
       run: |
-        python3 -m mypy --strict sansio_lsp_client
+        poetry run python3 -m mypy --strict sansio_lsp_client

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -13,7 +13,7 @@ jobs:
         python-version: 3.8
     - name: Install Dependencies
       run: |
-        pip install mypy
+        pip install poetry && poetry install
     - name: mypy
       run: |
         mypy --strict sansio_lsp_client

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -16,4 +16,4 @@ jobs:
         pip install poetry && poetry install
     - name: mypy
       run: |
-        mypy --strict sansio_lsp_client
+        python3 -m mypy --strict sansio_lsp_client

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -27,4 +27,5 @@ black = "^18.3-alpha.0"
 [tool.black]
 line-length = 79
 py36 = true
+# TODO: what's the purpose of specifying exclude here?
 exclude = 'build/|buck-out/|dist/|_build/|\.git/|\.hg/|\.mypy_cache/|\.tox/|\.venv/|/playground.py'

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -21,7 +21,7 @@ attrs = "^18.1"
 cattrs = "^0.9.0"
 
 [tool.poetry.dev-dependencies]
-mypy = "^0.620.0"
+mypy = "^0.782.0"
 black = "^18.3-alpha.0"
 
 [tool.black]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -26,6 +26,6 @@ black = "^18.3-alpha.0"
 
 [tool.black]
 line-length = 79
-py36 = true
+target-version = ["py36"]
 # TODO: what's the purpose of specifying exclude here?
 exclude = 'build/|buck-out/|dist/|_build/|\.git/|\.hg/|\.mypy_cache/|\.tox/|\.venv/|/playground.py'

--- a/sansio_lsp_client/__main__.py
+++ b/sansio_lsp_client/__main__.py
@@ -6,7 +6,6 @@ import socket
 import urllib.request
 
 from .client import Client
-from .errors import IncompleteResponseError
 from .events import Initialized, Shutdown, ShowMessageRequest, Completion
 from .structs import (
     CompletionTriggerKind,
@@ -36,14 +35,11 @@ def main() -> None:
     while True:
         sock.sendall(client.send())
 
-        try:
-            data = sock.recv(4096)
-            if not data:
-                break
-            events = list(client.recv(data))
-        except IncompleteResponseError as e:
-            continue
+        data = sock.recv(4096)
+        if not data:
+            break
 
+        events = list(client.recv(data))
         for event in events:
             if isinstance(event, Initialized):
                 print("Initialized!")
@@ -76,9 +72,16 @@ def main() -> None:
                 client.exit()
             elif isinstance(event, Completion):
                 print("Completions:")
-                pprint.pprint(
-                    [item.label for item in event.completion_list.items]
-                )
+                if event.completion_list is None:
+                    print("(None for whatever reason)")
+                elif isinstance(event.completion_list, list):
+                    pprint.pprint(
+                        [item.label for item in event.completion_list]
+                    )
+                else:
+                    pprint.pprint(
+                        [item.label for item in event.completion_list.items]
+                    )
 
                 client.did_close(
                     text_document=TextDocumentIdentifier(uri=file_uri)

--- a/sansio_lsp_client/client.py
+++ b/sansio_lsp_client/client.py
@@ -47,7 +47,10 @@ class ClientState(enum.Enum):
 class Client:
     # TODO: Save the encoding given here.
     def __init__(
-        self, process_id: int = None, root_uri: str = None, trace: str = "off"
+        self,
+        process_id: t.Optional[int] = None,
+        root_uri: t.Optional[str] = None,
+        trace: str = "off"
     ) -> None:
         self._state = ClientState.NOT_INITIALIZED
 
@@ -80,10 +83,10 @@ class Client:
         self._state = ClientState.WAITING_FOR_INITIALIZED
 
     @property
-    def state(self):
+    def state(self) -> ClientState:
         return self._state
 
-    def _send_request(self, method: str, params: JSONDict = None) -> int:
+    def _send_request(self, method: str, params: t.Optional[JSONDict] = None) -> int:
         id = self._id_counter
         self._id_counter += 1
 
@@ -93,11 +96,11 @@ class Client:
         )
         return id
 
-    def _send_notification(self, method: str, params: JSONDict = None) -> None:
+    def _send_notification(self, method: str, params: t.Optional[JSONDict] = None) -> None:
         self._send_buf += _make_request(method=method, params=params)
 
     def _send_response(
-        self, id: int, result: JSONDict = None, error: JSONDict = None
+        self, id: int, result: t.Optional[JSONDict] = None, error: t.Optional[JSONDict] = None
     ) -> None:
         self._send_buf += _make_response(id=id, result=result, error=error)
 
@@ -111,6 +114,7 @@ class Client:
         for message in messages:
             if isinstance(message, Response):
                 response = message
+                assert response.id is not None
                 request = self._unanswered_requests.pop(response.id)
 
                 # FIXME: The errors have meanings.
@@ -149,6 +153,7 @@ class Client:
 
                     events.append(Completion(response.id, completion_list))
                 elif request.method == "textDocument/willSaveWaitUntil":
+                    # FIXME: types don't match???
                     events.append(
                         WillSaveWaitUntilEdits(edits=response.result)
                     )
@@ -162,11 +167,12 @@ class Client:
                 def structure_request(event_cls: t.Type[E]) -> E:
                     if issubclass(event_cls, ServerRequest):
                         event = cattr.structure(request.params, event_cls)
+                        assert request.id is not None
                         event._id = request.id
                         event._client = self
-                        return event
+                        return t.cast('E', event)
                     elif issubclass(event_cls, ServerNotification):
-                        return cattr.structure(request.params, event_cls)
+                        return t.cast('E', cattr.structure(request.params, event_cls))
                     else:
                         raise TypeError(
                             "`event_cls` must be a subclass of ServerRequest or ServerNotification"
@@ -258,7 +264,7 @@ class Client:
         )
 
     def did_save(
-        self, text_document: TextDocumentIdentifier, text: str = None
+        self, text_document: TextDocumentIdentifier, text: t.Optional[str] = None
     ) -> None:
         assert self._state == ClientState.NORMAL
         params = {"textDocument": cattr.unstructure(text_document)}
@@ -276,7 +282,7 @@ class Client:
     def completions(
         self,
         text_document_position: TextDocumentPosition,
-        context: CompletionContext = None,
+        context: t.Optional[CompletionContext] = None,
     ) -> int:
         assert self._state == ClientState.NORMAL
         params = {}

--- a/sansio_lsp_client/client.py
+++ b/sansio_lsp_client/client.py
@@ -50,7 +50,7 @@ class Client:
         self,
         process_id: t.Optional[int] = None,
         root_uri: t.Optional[str] = None,
-        trace: str = "off"
+        trace: str = "off",
     ) -> None:
         self._state = ClientState.NOT_INITIALIZED
 
@@ -86,7 +86,9 @@ class Client:
     def state(self) -> ClientState:
         return self._state
 
-    def _send_request(self, method: str, params: t.Optional[JSONDict] = None) -> int:
+    def _send_request(
+        self, method: str, params: t.Optional[JSONDict] = None
+    ) -> int:
         id = self._id_counter
         self._id_counter += 1
 
@@ -96,11 +98,16 @@ class Client:
         )
         return id
 
-    def _send_notification(self, method: str, params: t.Optional[JSONDict] = None) -> None:
+    def _send_notification(
+        self, method: str, params: t.Optional[JSONDict] = None
+    ) -> None:
         self._send_buf += _make_request(method=method, params=params)
 
     def _send_response(
-        self, id: int, result: t.Optional[JSONDict] = None, error: t.Optional[JSONDict] = None
+        self,
+        id: int,
+        result: t.Optional[JSONDict] = None,
+        error: t.Optional[JSONDict] = None,
     ) -> None:
         self._send_buf += _make_response(id=id, result=result, error=error)
 
@@ -170,9 +177,11 @@ class Client:
                         assert request.id is not None
                         event._id = request.id
                         event._client = self
-                        return t.cast('E', event)
+                        return t.cast("E", event)
                     elif issubclass(event_cls, ServerNotification):
-                        return t.cast('E', cattr.structure(request.params, event_cls))
+                        return t.cast(
+                            "E", cattr.structure(request.params, event_cls)
+                        )
                     else:
                         raise TypeError(
                             "`event_cls` must be a subclass of ServerRequest or ServerNotification"
@@ -264,7 +273,9 @@ class Client:
         )
 
     def did_save(
-        self, text_document: TextDocumentIdentifier, text: t.Optional[str] = None
+        self,
+        text_document: TextDocumentIdentifier,
+        text: t.Optional[str] = None,
     ) -> None:
         assert self._state == ClientState.NORMAL
         params = {"textDocument": cattr.unstructure(text_document)}
@@ -290,4 +301,5 @@ class Client:
         if context is not None:
             params.update(cattr.unstructure(context))
         return self._send_request(
-            method="textDocument/completion", params=params)
+            method="textDocument/completion", params=params
+        )

--- a/sansio_lsp_client/client.py
+++ b/sansio_lsp_client/client.py
@@ -163,7 +163,9 @@ class Client:
                 elif request.method == "textDocument/willSaveWaitUntil":
                     # FIXME: types don't match???
                     events.append(
-                        WillSaveWaitUntilEdits(edits=t.cast(t.List[TextEdit], response.result))
+                        WillSaveWaitUntilEdits(
+                            edits=t.cast(t.List[TextEdit], response.result)
+                        )
                     )
                 else:
                     raise NotImplementedError((response, request))

--- a/sansio_lsp_client/client.py
+++ b/sansio_lsp_client/client.py
@@ -31,6 +31,7 @@ from .structs import (
     VersionedTextDocumentIdentifier,
     TextDocumentContentChangeEvent,
     TextDocumentSaveReason,
+    TextEdit,
 )
 from .io_handler import _make_request, _parse_messages, _make_response
 
@@ -162,7 +163,7 @@ class Client:
                 elif request.method == "textDocument/willSaveWaitUntil":
                     # FIXME: types don't match???
                     events.append(
-                        WillSaveWaitUntilEdits(edits=response.result)
+                        WillSaveWaitUntilEdits(edits=t.cast(t.List[TextEdit], response.result))
                     )
                 else:
                     raise NotImplementedError((response, request))

--- a/sansio_lsp_client/events.py
+++ b/sansio_lsp_client/events.py
@@ -24,7 +24,7 @@ class Event:
 
 @attrs
 class ServerRequest(Event):
-    _client: 'Client' = attrib(init=False)
+    _client: "Client" = attrib(init=False)
     _id: int = attrib(init=False)
 
 

--- a/sansio_lsp_client/events.py
+++ b/sansio_lsp_client/events.py
@@ -3,6 +3,8 @@ import typing as t
 import attr
 from attr import attrs, attrib
 
+if t.TYPE_CHECKING:  # avoid import cycle at runtime
+    from .client import Client
 from .structs import (
     JSONDict,
     Diagnostic,
@@ -22,7 +24,7 @@ class Event:
 
 @attrs
 class ServerRequest(Event):
-    _client: "Client" = attrib(init=False)
+    _client: 'Client' = attrib(init=False)
     _id: int = attrib(init=False)
 
 
@@ -53,7 +55,7 @@ class ShowMessageRequest(ServerRequest):
     message: str = attrib()
     actions: t.Optional[t.List[MessageActionItem]] = attrib()
 
-    def reply(self, action: MessageActionItem = None) -> None:
+    def reply(self, action: t.Optional[MessageActionItem] = None) -> None:
         """
         Reply to the ShowMessageRequest with the user's selection.
 

--- a/sansio_lsp_client/io_handler.py
+++ b/sansio_lsp_client/io_handler.py
@@ -99,7 +99,7 @@ def _make_response(
 # _parse_one_message returns None when there are no more messages, and an empty
 # iterator when a message was parsed but no things were created.
 def _parse_one_message(
-    response_buf: bytearray
+    response_buf: bytearray,
 ) -> t.Optional[t.Iterable[t.Union[Response, Request]]]:
     if b"\r\n\r\n" not in response_buf:
         return None
@@ -183,7 +183,7 @@ def _parse_one_message(
 
 
 def _parse_messages(
-    response_buf: bytearray
+    response_buf: bytearray,
 ) -> t.Iterator[t.Union[Response, Request]]:
     while True:
         parsed = _parse_one_message(response_buf)

--- a/sansio_lsp_client/io_handler.py
+++ b/sansio_lsp_client/io_handler.py
@@ -98,7 +98,9 @@ def _make_response(
 #
 # _parse_one_message returns None when there are no more messages, and an empty
 # iterator when a message was parsed but no things were created.
-def _parse_one_message(response_buf: bytearray) -> t.Optional[t.Iterable[t.Union[Response, Request]]]:
+def _parse_one_message(
+    response_buf: bytearray
+) -> t.Optional[t.Iterable[t.Union[Response, Request]]]:
     if b"\r\n\r\n" not in response_buf:
         return None
 
@@ -138,14 +140,13 @@ def _parse_one_message(response_buf: bytearray) -> t.Optional[t.Iterable[t.Union
     raw_content = raw_content[:content_length]
     unused_bytes_count = len(raw_content[content_length:])
 
-
     # This is a good place for deleting unnecessary stuff from response_buf
     # because if the code below fails, then leaving the cause of failure to
     # response_buf would cause this function to fail every time in the future
     # when called with the same response_buf. I think I've had this issue a
     # long time ago, and it was annoying how one response parsing error would
     # also block the parsing of any future responses.
-    if unused_bytes_count == 0:     # 'del response_buf[:-0]' does wrong thing
+    if unused_bytes_count == 0:  # 'del response_buf[:-0]' does wrong thing
         response_buf.clear()
     else:
         del response_buf[:-unused_bytes_count]
@@ -181,7 +182,9 @@ def _parse_one_message(response_buf: bytearray) -> t.Optional[t.Iterable[t.Union
         return [do_it(content)]
 
 
-def _parse_messages(response_buf: bytearray) -> t.Iterator[t.Union[Response, Request]]:
+def _parse_messages(
+    response_buf: bytearray
+) -> t.Iterator[t.Union[Response, Request]]:
     while True:
         parsed = _parse_one_message(response_buf)
         if parsed is None:

--- a/sansio_lsp_client/io_handler.py
+++ b/sansio_lsp_client/io_handler.py
@@ -21,8 +21,8 @@ def _make_headers(content_length: int, encoding: str = "utf-8") -> bytes:
 
 def _make_request(
     method: str,
-    params: JSONDict = None,
-    id: int = None,
+    params: t.Optional[JSONDict] = None,
+    id: t.Optional[int] = None,
     *,
     encoding: str = "utf-8",
 ) -> bytes:
@@ -49,8 +49,8 @@ def _make_request(
 
 def _make_response(
     id: int,
-    result: JSONDict = None,
-    error: JSONDict = None,
+    result: t.Optional[JSONDict] = None,
+    error: t.Optional[JSONDict] = None,
     *,
     encoding: str = "utf-8",
 ) -> bytes:
@@ -159,12 +159,14 @@ def _parse_one_message(response_buf: bytearray) -> t.Optional[t.Iterable[t.Union
         # filled in with None)
         try:
             request = cattr.structure(data, Request)
+            assert isinstance(request, Request)
             return request
         except TypeError:
             pass
 
         try:
             response = cattr.structure(data, Response)
+            assert isinstance(response, Response)
             return response
         except TypeError:
             pass

